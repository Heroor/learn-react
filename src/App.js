import React, { Component } from 'react';
// import {
//   BrowserRouter as Router,
//   Route,
//   Link
// } from 'react-router-dom'
import './App.css';
import {
  storngLog,
  dangerLog,
  successLog,
  errorLog
} from './config/colorLog'

import Search from './components/Search'
import Table from './components/Table'
import Button from './components/Button'



const DEFAULT_QUERY = 'react'
const DEFAULT_HPP = '5'
const PATH_BASE = 'https://hn.algolia.com/api/v1'
const PATH_SEARCH = '/search'
const PARAM_SEARCH = 'query='
const PARAM_PAGE = 'page='
const PARAM_HPP = 'hitsPerPage='


// root component of webApp
class App extends Component {
  // called before component will mount
  constructor(props) {
    // component init
    console.log('%c[constructor function]', storngLog)
    super(props)
    // here we can init state or methods
    this.state = {
      searchValue: DEFAULT_QUERY,
      searchKey: '',
      results: null,
      error: null
    }
    // bind function to the this(instance)
    this.remove = this.remove.bind(this)
    this.onInputChange = this.onInputChange.bind(this)
    this.onSearchSubmit = this.onSearchSubmit.bind(this)
    this.setSearchTopStories = this.setSearchTopStories.bind(this)
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this)
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this)
  }

  // class's methods
  remove(item) {
    const {searchKey, results} = this.state
    const {hits, page} = results[searchKey]

    const isNotItem = v => v !== item
    const updateHits = hits.filter(isNotItem)
    this.setState({
      results: {
        ...results,
        [searchKey]: {
          hits: updateHits,
          page
        }
      }
    })
  }

  onInputChange(event) {
    this.setState({searchValue: event.target.value})
  }

  onSearchSubmit(event) {
    const {searchValue} = this.state
    this.setState({
      searchKey: searchValue
    })
    this.needsToSearchTopStories(searchValue) && this.fetchSearchTopStories(searchValue)

    event.preventDefault()
  }

  setSearchTopStories(result) {
    const {hits, page} = result
    const {searchKey, results} = this.state

    const oldHits = results && results[searchKey] ?
      results[searchKey].hits :
      []
    const updateHits = [
      ...oldHits,
      ...hits
    ]
    this.setState({
      results: {
        ...results,
        [searchKey]: {
          hits: updateHits,
          page
        }
      }
    })
  }

  needsToSearchTopStories(searchKey) {
    return !this.state.results[searchKey]
  }

  fetchSearchTopStories(searchValue, page = 0) {
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchValue}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(res => res.json())
      .then(this.setSearchTopStories)
      .then(res => this.setState({
        error: null
      }))
      .catch(e => this.setState({
        error: e
      }))
  }

  /*------------------------component lifecycle-------------------------*/

    /* [component mount order]:
     * constructor()
     * componentWillMount()
     * render()
     * componentDidMount()
     */

    /* [component update order]:
     * componentWillReceiveProps()
     * shouldComponentUpdate()
     * componentWillUpdate()
     * render()
     * componentDidUpdate(
     */

     /* [component unmount]
      * componentWillUnmount()
      */

  /*---------------------------------------------------------------------*/

  // lifecycle hooks
  componentWillMount() {
    // called before rendering the component
    console.log('%c[componentWillMount hook]', dangerLog)
  }

  componentDidMount() {
    // called after the component is mounted
    console.log('%c[componentDidMount hook]', dangerLog)
    // here we can [ftech API, change state, ...]

    const {searchValue} = this.state
    this.setState({
      searchKey: searchValue
    })
    this.fetchSearchTopStories(searchValue)
  }

  componentWillReceiveProps(nextProps) {
    console.log('[componentWillReceiveProps hook]')
  }

  shouldComponentUpdate(nextProps, nextState) {
    // called each time the component state or props changes
    // called before componentWillUpdate
    console.log('[shouldComponentUpdate hook]')
    // here we can [Performance optimization, ... ]

    // must return a Boolean value:
    // true - component will be rendered
    // false - component dose not rendered
    return true
  }

  componentWillUpdate(nextProps, nextState) {
    // called before the component render
    console.log('%c[componentWillUpdate hook]', successLog)
    // here we can handle state or props before render
    // here can not use 'this.setState', it will infinite loop

    // nextState.list.splice(-5)
  }

  componentDidUpdate(prevProps, prevState) {
    // called after the component render
    console.log('%c[componentDidUpdate hook]', successLog)
    // here we can [handle DOM, handle async, ...]
    // here can not use 'this.setState', it will infinite loop

    // console.log(document.querySelector('.page ul'))
  }

  componentWillUnmount() {
    // called before the component unmount
    console.log('%c[componentWillUnmount hook]', errorLog)
  }

  componentDidCatch(error, info) {
    // Catch component error
    console.log(error, info)
  }

  // The element returned by the component
  // called after the component will mount or
  // after the component will update(state or props have changed)
  render() {
    console.log('%c[render function]', storngLog)
    const {
      searchValue,
      searchKey,
      results,
      error
    } = this.state

    const page = (
      results &&
      results[searchKey] &&
      results[searchKey].page
    ) || 0

    const list = (
      results &&
      results[searchKey] &&
      results[searchKey].hits
     ) || []

    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchValue}
            onChange={this.onInputChange}
            onSubmit={this.onSearchSubmit}
          >
            search
          </Search>
        </div>
        {
          error ?
          <div className="interactions">
            <p>Oops! something went wrong!</p>
          </div> :
          <Table
            list={list}
            remove={this.remove}
          />
        }
        <div className="interactions">
          <Button
            onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}
          >
            more
          </Button>
        </div>
      </div>
    );
  }
}

export default App;
