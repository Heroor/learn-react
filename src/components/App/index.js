import React, { Component } from 'react'
import fetch from 'isomorphic-fetch'
// import {
//   BrowserRouter as Router,
//   Route,
//   Link
// } from 'react-router-dom'
import './index.css'
import {
  storngLog,
  dangerLog,
  successLog,
  errorLog
} from '../../config/colorLog'

// components
import Search from '../Search'
import Table from '../Table'
import Button from '../Button'
import {withLoading} from '../Loading'
import logoImg from '../../logo.svg'

// const
import {
  DEFAULT_QUERY,
  DEFAULT_HPP,
  PATH_BASE,
  PATH_SEARCH,
  PARAM_SEARCH,
  PARAM_PAGE,
  PARAM_HPP
} from '../../constants'


// root component of webApp
class App extends Component {
  // called before component will mount
  constructor(props) {
    // component init
    console.log('%c[constructor function]', storngLog)
    super(props)
    // here we can init state or methods
    this.state = {
      searchValue: DEFAULT_QUERY,
      searchKey: '',
      results: null,
      error: null,
      isLoading: false
    }
    // bind function to the this(instance)
    this.onRemove = this.onRemove.bind(this)
    this.onInputChange = this.onInputChange.bind(this)
    this.onSearchSubmit = this.onSearchSubmit.bind(this)
    this.setSearchTopStories = this.setSearchTopStories.bind(this)
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this)
    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this)
  }

  // class's methods
  onRemove(item) {
    this.setState(updateOnRemove(item))
  }

  onInputChange(event) {
    this.setState({ searchValue: event.target.value })
  }

  onSearchSubmit(event) {
    const { searchValue } = this.state
    this.setState({
      searchKey: searchValue
    })
    !this.state.isLoading && this.needsToSearchTopStories(searchValue) && this.fetchSearchTopStories(searchValue)

    event.preventDefault()
  }

  setSearchTopStories(result) {
    const { hits, page } = result
    this.setState(updateSearchTopStoriesState(hits, page))
  }

  needsToSearchTopStories(searchKey) {
    return !this.state.results[searchKey]
  }

  fetchSearchTopStories(searchValue, page = 0) {
    this.setState({
      isLoading: true
    })
    fetch(`${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchValue}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HPP}`)
      .then(res => res.json())
      .then(this.setSearchTopStories)
      .then(res => this.setState({
        error: null
      }))
      .catch(e => this.setState({
        error: e
      }))
  }

  /*-----------------component lifecycle-------------------*/

    /* [component mountimg]:
     * constructor()
     * componentWillMount()
     * render()
     * componentDidMount()
     */

    /* [component updating]:
     * componentWillReceiveProps()
     * shouldComponentUpdate()
     * componentWillUpdate()
     * render()
     * componentDidUpdate(
     */

     /* [component unmounting]
      * componentWillUnmount()
      */

  /*---------------------------------------------------------*/

  // lifecycle hooks
  componentWillMount() {
    // called before rendering the component
    console.log('%c[componentWillMount hook]', dangerLog)
  }

  componentDidMount() {
    // called after the component is mounted
    console.log('%c[componentDidMount hook]', dangerLog)
    // here we can [ftech API, change state, ...]

    const { searchValue } = this.state
    this.setState({
      searchKey: searchValue
    })
    this.fetchSearchTopStories(searchValue)
  }

  componentWillReceiveProps(nextProps) {
    // called when the props received by the component have changed
    console.log('[componentWillReceiveProps hook]')
  }

  shouldComponentUpdate(nextProps, nextState) {
    // called each time the component state or props changes
    // called before componentWillUpdate
    console.log('[shouldComponentUpdate hook]')
    // here we can [Performance optimization, ... ]

    // must return a Boolean value:
    // true - component will be rendered
    // false - component dose not rendered
    return true
  }

  componentWillUpdate(nextProps, nextState) {
    // called before the component render
    console.log('%c[componentWillUpdate hook]', successLog)
    // here we can handle state or props before render
    // here can not use 'this.setState', it will infinite loop

    // nextState.list.splice(-5)
  }

  componentDidUpdate(prevProps, prevState) {
    // called after the component render
    console.log('%c[componentDidUpdate hook]', successLog)
    // here we can [handle DOM, handle async, ...]
    // here can not use 'this.setState', it will infinite loop

    // console.log(document.querySelector('.page ul'))
  }

  componentWillUnmount() {
    // called before the component unmount
    console.log('%c[componentWillUnmount hook]', errorLog)
  }

  componentDidCatch(error, info) {
    // Catch component error
    console.log(error, info)
  }

  // The element returned by the component
  // called after the component will mount or
  // after the component will update(state or props have changed)
  render() {
    console.log('%c[render function]', storngLog)
    const {
      searchValue,
      searchKey,
      results,
      error,
      isLoading
    } = this.state

    const page = (
      results &&
      results[searchKey] &&
      results[searchKey].page
    ) || 0

    const list = (
      results &&
      results[searchKey] &&
      results[searchKey].hits
    ) || []

    return (
      <div className="page">
        <div className="interactions">
          <Search
            value={searchValue}
            isLoading={isLoading}
            onChange={this.onInputChange}
            onSubmit={this.onSearchSubmit}
          >
            search
          </Search>
        </div>
        {/*
          error ?
          <div className="interactions">
            <p>Oops! something went wrong!</p>
          </div> :
          <Table
            list={list}
            remove={this.onRemove}
          />
        */}
        <TableWithError 
            isError={error}
            list={list}
            remove={this.onRemove}
          />
        <div className="interactions">
          <ButtonWithLoading
            isLoading={isLoading}
            loadingChildren={
              <img src={logoImg} alt="loading"/>
            }
            onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}
          >
            More
          </ButtonWithLoading>
        </div>
      </div>
    )
  }
}


const WithError = Component => ({isError, ...rest}) => (
  isError? 
  <div className="interactions">
    <p>Oops! something went wrong!</p>
  </div> :
  <Component {...rest}/>
)

const TableWithError = WithError(Table)
const ButtonWithLoading = withLoading(Button)

export const updateSearchTopStoriesState = (hits, page) =>
  preState => {
    const { searchKey, results } = preState

    const oldHits = results && results[searchKey] ?
      results[searchKey].hits : []

    const updateHits = [
      ...oldHits,
      ...hits
    ]

    return {
      results: {
        ...results,
        [searchKey]: {
          hits: updateHits,
          page
        }
      },
      isLoading: false
    }
  }

export const updateOnRemove = item => prevState => {
  const {
    searchKey,
    results
  } = prevState
  const { hits, page } = results[searchKey]
  const isNotItem = v => v !== item
  const updateHits = hits.filter(isNotItem)
  return {
    results: {
      ...results,
      [searchKey]: {
        hits: updateHits,
        page
      }
    }
  }
}

export default App

// replaceState
// forceUpdate

/* todo :
 * flow
 */