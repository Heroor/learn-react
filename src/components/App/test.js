// 组件测试
// $ npm test

import React from 'react'
import ReactDOM from 'react-dom'
import App, {
  updateSearchTopStoriesState
} from './index.js'
import renderer from 'react-test-renderer'
import { hitList } from '../../config/TestConfig'

// App test
describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render( <App/> , div)
  })

  // 创建测试快照
  // i. 快照可以保存上次测试（如模拟DOM渲染）的结果，下次的测试会与已保存的快照作对比，来找出发生变化的地方并报告错误，保证程序不被意外改变；
  // ii. 如果最新的测试结果变化了并且是预期的正确行为，可以手动保存新快照来更新旧快照
  test('has a valid snashot', () => {
    // 模拟渲染组件的DOM
    const component = renderer.create(
      <App />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

// updateSearchTopStoriesState
describe('updateSearchTopStoriesState', () => {
  it('updateSearchTopStoriesState is running', () => {
    updateSearchTopStoriesState(hitList, 0)
  })
})
