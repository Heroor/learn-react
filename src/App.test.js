// 组件测试
// $ npm test

import React from 'react'
import ReactDOM from 'react-dom'
import App, {
  Search,
  Table,
  Button
} from './App'
import renderer from 'react-test-renderer'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

Enzyme.configure({
  adapter: new Adapter()
})

// App test
describe('App', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render( <App/> , div)
  })

  // 创建测试快照
  // i. 快照可以保存上次测试（如模拟DOM渲染）的结果，下次的测试会与已保存的快照作对比，来找出发生变化的地方并报告错误，保证程序不被意外改变；
  // ii. 如果最新的测试结果变化了并且是预期的正确行为，可以手动保存新快照来更新旧快照
  test('has a valid snashot', () => {
    // 模拟渲染组件的DOM
    const component = renderer.create(
      <App />
    )
    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})

// Search test
describe('Search', () => {
  const props = {
    value: 'ipt val',
    onChange() {
      console.log('onChange')
    },
    onSubmit() {
      console.log('onSubmit')
    },
  }
  it('renders withou crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Search {...props}>{props.children}</Search>, div)
  })
  test('has a valid snapshot', () => {
    renderComponentToSnapshot(
      <Search {...props}>{props.children}</Search>
    )
  })
  test('has a valid text snapshot', () => {
    let text = 'Lorem ipsum dolor sit amet quisquam.'
    expect(text).toMatchSnapshot()
  })
})

// Button test
describe('Button', () => {
  const props = {
    onClick() {
      console.log('click!')
    },
    children: 'More'
  }
  it('render without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Button {...props}>{props.children}</Button>, div)
  })
  test('has a valid shapshot', () => {
    renderComponentToSnapshot(
      <Button {...props}>{props.children}</Button>
    )
  })
})

// Table test
describe('Table', () => {
  const props = {
    list: [
      { title: '1', author: '1', num_comments: 1, point: 1, objectID: 'y' },
      { title: '2', author: '2', num_comments: 2, point: 2, objectID: 'x' }
    ],
    remove() {
      console.log('remove!')
    }
  }

  it('render without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<Table {...props} />, div)
  })

  test('has a calid snapshot', () => {
    renderComponentToSnapshot(<Table {...props} />)
  })

  // 单元测试
  //Enzyme API 中总共有三种渲染机制。你已经知道了 shallow() ，这里还有 mount() 和
  // render() 方法。这两种方式都会初始化父组件和所有的子组件。此外 mount() 还给予
  // 你调用组件生命周期的方法。但是什时候该使用哪种渲染机制呢？这里有一些建议：
  // • 不论怎样都优先尝试使用浅渲染（ shallow() ）
  // • 如果需要测试 componentDidMount() 或 componentDidUpdate() ，使用 mount()
  // • 如果你想测试组件的生命周期和子组件的行为，使用 mount()
  // • 如果你想测试一个组件的子组件的渲染，并且不关心生命周期方法和减少些渲染的
  // 花销的话，使用 render()
  it('shows two items in list', () => {
    const element = mount(
      <Table {...props} />
    )
    expect(element.find('.table-row').length).toBe(2)
  })
})


function renderComponentToSnapshot (DOM) {
  var component = renderer.create(DOM)
  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
}


