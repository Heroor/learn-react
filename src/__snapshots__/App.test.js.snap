// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snashot 1`] = `
<div
  className="page"
>
  <div
    className="interactions"
  >
    <form
      onSubmit={[Function]}
    >
      <input
        onChange={[Function]}
        type="text"
        value="redux"
      />
      <button
        type="submit"
      >
        search
      </button>
    </form>
  </div>
  <ul
    className="table"
  />
  <div
    className="interactions"
  >
    <button
      className=""
      onClick={[Function]}
      type="button"
    >
      more
    </button>
  </div>
</div>
`;

exports[`Button has a valid shapshot 1`] = `
<button
  className=""
  onClick={[Function]}
  type="button"
>
  More
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    type="text"
    value="ipt val"
  />
  <button
    type="submit"
  >
    search...
  </button>
</form>
`;

exports[`Search has a valid text snapshot 1`] = `"Lorem ipsum dolor sit amet quisquam."`;

exports[`Table has a calid snapshot 1`] = `
<ul
  className="table"
>
  <li
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    />
    <span
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <a
        href={undefined}
        target="blank"
      >
        1
      </a>
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      1
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        [DEL]
      </button>
    </span>
  </li>
  <li
    className="table-row"
  >
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    />
    <span
      style={
        Object {
          "width": "50%",
        }
      }
    >
      <a
        href={undefined}
        target="blank"
      >
        2
      </a>
    </span>
    <span
      style={
        Object {
          "width": "20%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      2
    </span>
    <span
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <button
        className="button-inline"
        onClick={[Function]}
        type="button"
      >
        [DEL]
      </button>
    </span>
  </li>
</ul>
`;
